core/commonTest/src/kotlinx/serialization/SerializersModuleTest.kt

```diff
@@ -0,0 +1,116 @@
+
+class SerializersModuleTest {
+    @Serializable
+    object Object
+
+    @Serializable
+    sealed class SealedParent {
+        @Serializable
+        data class Child(val i: Int) : SealedParent()
+    }
+
+    @Serializable
+    abstract class Abstract
+
+    @Serializable
+    enum class SerializableEnum { A, B }
+
+    @Serializable(CustomSerializer::class)
+    class WithCustomSerializer(val i: Int)
+
+    @Serializer(forClass = WithCustomSerializer::class)
+    object CustomSerializer
+
+    @Serializable
+    class Parametrized<T : Any>(val a: T)
+
+    class ContextualType(val i: Int)
+
+    @Serializer(forClass = ContextualType::class)
+    object ContextualSerializer
+
+    class FileContextualType(val i: Int)
+
+    @Serializer(forClass = FileContextualType::class)
+    object FileContextualSerializer
+
+    @Serializable
+    class ContextualHolder(@Contextual val contextual: ContextualType, val fileContextual: FileContextualType)
+
+    @Test
+    fun testCompiled() = noJsLegacy {
+        assertSame<KSerializer<*>>(Object.serializer(), serializer(Object::class, emptyList(), false))
+        assertSame<KSerializer<*>>(SealedParent.serializer(), serializer(SealedParent::class, emptyList(), false))
+        assertSame<KSerializer<*>>(
+            SealedParent.Child.serializer(),
+            serializer(SealedParent.Child::class, emptyList(), false)
+        )
+
+        assertSame<KSerializer<*>>(Abstract.serializer(), serializer(Abstract::class, emptyList(), false))
+        assertSame<KSerializer<*>>(SerializableEnum.serializer(), serializer(SerializableEnum::class, emptyList(), false))
+    }
+
+    @Test
+    fun testBuiltIn() {
+        assertSame<KSerializer<*>>(Int.serializer(), serializer(Int::class, emptyList(), false))
+    }
+
+    @Test
+    fun testCustom() {
+        val m = SerializersModule { }
+        assertSame<KSerializer<*>>(CustomSerializer, m.serializer(WithCustomSerializer::class, emptyList(), false))
+    }
+
+    @Test
+    fun testParametrized() {
+        val serializer = serializer(Parametrized::class, listOf(Int.serializer()), false)
+        assertEquals<KClass<*>>(Parametrized.serializer(Int.serializer())::class, serializer::class)
+        assertEquals(PrimitiveKind.INT, serializer.descriptor.getElementDescriptor(0).kind)
+
+        val mapSerializer = serializer(Map::class, listOf(String.serializer(), Int.serializer()), false)
+        assertIs<MapLikeSerializer<*, *, *, *>>(mapSerializer)
+        assertEquals(PrimitiveKind.STRING, mapSerializer.descriptor.getElementDescriptor(0).kind)
+        assertEquals(PrimitiveKind.INT, mapSerializer.descriptor.getElementDescriptor(1).kind)
+    }
+
+    @Test
+    fun testUnsupportedArray() {
+        assertFails {
+            serializer(Array::class, listOf(Int.serializer()), false)
+        }
+    }
+
+    @Test
+    fun testContextual() {
+        val m = SerializersModule {
+            contextual<FileContextualType>(FileContextualSerializer)
```

Comment:
It doesn't matter if there is `@file:UseContextualSerialization` or not, because it is a reflective lookup unaffected by pluing. It will always try to search something in the context

Reply:
This is exactly a test for runtime, that there is a visit to this search branch. Perhaps it is redundant

Reply:
I mean, there should be no difference in code path for `FileContextualType` and `ContextualType`